File: whoop_sync/calendar_client.py
from typing import Optional
from urllib.parse import quote

import requests

CALENDAR_API_BASE = "https://www.googleapis.com/calendar/v3"
GOOGLE_TOKEN_URL = "https://oauth2.googleapis.com/token"


def exchange_code(client_id: str, client_secret: str, code: str, redirect_uri: str) -> dict:
    res = requests.post(
        GOOGLE_TOKEN_URL,
        data={
            "client_id": client_id,
            "client_secret": client_secret,
            "code": code,
            "redirect_uri": redirect_uri,
            "grant_type": "authorization_code",
        },
    )
    res.raise_for_status()
    return res.json()


def get_access_token(client_id: str, client_secret: str, refresh_token: str) -> str:
    """Google refresh tokens don't rotate, so callers just keep the access token for one run."""
    res = requests.post(
        GOOGLE_TOKEN_URL,
        data={
            "client_id": client_id,
            "client_secret": client_secret,
            "refresh_token": refresh_token,
            "grant_type": "refresh_token",
        },
    )
    res.raise_for_status()
    return res.json()["access_token"]


def upsert_event(access_token: str, calendar_id: str, event: dict) -> None:
    """
    Idempotent create-or-update keyed on extendedProperties.private.whoopRecordId,
    so the Worker's webhook handler and this reconciliation job can safely
    process the same WHOOP record without creating duplicate events.
    """
    whoop_record_id = event["extendedProperties"]["private"]["whoopRecordId"]
    existing_id = _find_event_id(access_token, calendar_id, whoop_record_id)
    encoded_calendar_id = quote(calendar_id, safe="")
    headers = {"Authorization": f"Bearer {access_token}", "Content-Type": "application/json"}

    if existing_id:
        url = f"{CALENDAR_API_BASE}/calendars/{encoded_calendar_id}/events/{existing_id}"
        res = requests.patch(url, headers=headers, json=event)
    else:
        url = f"{CALENDAR_API_BASE}/calendars/{encoded_calendar_id}/events"
        res = requests.post(url, headers=headers, json=event)
    res.raise_for_status()


def _find_event_id(access_token: str, calendar_id: str, whoop_record_id: str) -> Optional[str]:
    encoded_calendar_id = quote(calendar_id, safe="")
    res = requests.get(
        f"{CALENDAR_API_BASE}/calendars/{encoded_calendar_id}/events",
        headers={"Authorization": f"Bearer {access_token}"},
        params={"privateExtendedProperty": f"whoopRecordId={whoop_record_id}"},
    )
    res.raise_for_status()
    items = res.json().get("items", [])
    return items[0]["id"] if items else None

File: shared/calendar-client.ts
const CALENDAR_API_BASE = "https://www.googleapis.com/calendar/v3";
const GOOGLE_TOKEN_URL = "https://oauth2.googleapis.com/token";

export interface CalendarEvent {
  summary: string;
  description: string;
  start: { dateTime: string };
  end: { dateTime: string };
  colorId: string;
  extendedProperties: { private: Record<string, string> };
}

export interface GoogleCredentials {
  clientId: string;
  clientSecret: string;
}

/** Google refresh tokens don't rotate, so callers just cache the access token in memory per run. */
export async function getGoogleAccessToken(creds: GoogleCredentials, refreshToken: string): Promise<string> {
  const res = await fetch(GOOGLE_TOKEN_URL, {
    method: "POST",
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
    body: new URLSearchParams({
      client_id: creds.clientId,
      client_secret: creds.clientSecret,
      refresh_token: refreshToken,
      grant_type: "refresh_token",
    }),
  });
  if (!res.ok) {
    throw new Error(`Google token refresh failed: ${res.status} ${await res.text()}`);
  }
  const body = (await res.json()) as { access_token: string };
  return body.access_token;
}

/**
 * Idempotent create-or-update keyed on extendedProperties.private.whoopRecordId,
 * so the webhook handler and the daily reconciliation job can safely process
 * the same WHOOP record without creating duplicate events.
 */
export async function upsertEvent(
  accessToken: string,
  calendarId: string,
  event: CalendarEvent,
): Promise<void> {
  const whoopRecordId = event.extendedProperties.private.whoopRecordId;
  const existingId = await findEventIdByWhoopRecordId(accessToken, calendarId, whoopRecordId);

  const url = existingId
    ? `${CALENDAR_API_BASE}/calendars/${encodeURIComponent(calendarId)}/events/${existingId}`
    : `${CALENDAR_API_BASE}/calendars/${encodeURIComponent(calendarId)}/events`;

  const res = await fetch(url, {
    method: existingId ? "PATCH" : "POST",
    headers: {
      Authorization: `Bearer ${accessToken}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(event),
  });
  if (!res.ok) {
    throw new Error(`Calendar upsert failed: ${res.status} ${await res.text()}`);
  }
}

async function findEventIdByWhoopRecordId(
  accessToken: string,
  calendarId: string,
  whoopRecordId: string,
): Promise<string | null> {
  const params = new URLSearchParams({
    privateExtendedProperty: `whoopRecordId=${whoopRecordId}`,
  });
  const res = await fetch(`${CALENDAR_API_BASE}/calendars/${encodeURIComponent(calendarId)}/events?${params}`, {
    headers: { Authorization: `Bearer ${accessToken}` },
  });
  if (!res.ok) {
    throw new Error(`Calendar lookup failed: ${res.status} ${await res.text()}`);
  }
  const body = (await res.json()) as { items: { id: string }[] };
  return body.items[0]?.id ?? null;
}
